### PR TITLE
Issue47 Maintain symbolic expressions containing error for path conditions

### PR DESCRIPTION
**Pull request by Himeshi**

### CODE REVIEW

File: lib/Core/ErrorState.cpp [hunk 1]
```diff
@@ -332,8 +339,74 @@ ref<Expr> ErrorState::propagateError(Executor *executor,
...
+    llvm::ICmpInst *ii = cast<llvm::ICmpInst>(ci);
+
+    //(1 - ex)
+    ref<Expr> lError = SubExpr::create(ConstantExpr::create(1, Expr::Int8),
+                                       arguments.at(0).error);
+    ref<Expr> rError = SubExpr::create(ConstantExpr::create(1, Expr::Int8),
+                                       arguments.at(1).error);
+
+    // x * (1 - ex)
+    ref<Expr> leftMul = MulExpr::create(arguments.at(0).error, lError);
```
Comment (domainexpert) [inline lib/Core/ErrorState.cpp]: I think here it should be arguments.at(0).value.

File: lib/Core/ErrorState.cpp [hunk 2]
```diff
@@ -332,8 +339,74 @@ ref<Expr> ErrorState::propagateError(Executor *executor,
...
+    ref<Expr> rightMul = MulExpr::create(arguments.at(1).error, rError);
```
Comment (domainexpert) [inline lib/Core/ErrorState.cpp]: I think here it should be arguments.at(1).value.

File: lib/Core/SeedInfo.cpp [hunk 3]
```diff
@@ -105,11 +105,11 @@ void SeedInfo::patchSeed(const ExecutionState &state,
         assert(success && "FIXME: Unhandled solver failure");            
         (void) success;
         it2->second[i] = value->getZExtValue(8);
-        tmp.addConstraint(EqExpr::create(read, 
-                                         ConstantExpr::alloc(it2->second[i], 
-                                                             Expr::Int8)));
+        tmp.addConstraint(EqExpr::create(read, ConstantExpr::alloc(
+                                                   it2->second[i], Expr::Int8)),
+                          ConstantExpr::create(1, Expr::Int8));
```
Comment (domainexpert) [inline lib/Core/SeedInfo.cpp]: Why is the error amount here 1?

File: lib/Core/SeedInfo.cpp [hunk 4]
```diff
@@ -142,11 +142,11 @@ void SeedInfo::patchSeed(const ExecutionState &state,
         assert(success && "FIXME: Unhandled solver failure");            
         (void) success;
         it->second[i] = value->getZExtValue(8);
-        tmp.addConstraint(EqExpr::create(read, 
-                                         ConstantExpr::alloc(it->second[i], 
-                                                             Expr::Int8)));
+        tmp.addConstraint(EqExpr::create(read, ConstantExpr::alloc(
+                                                   it->second[i], Expr::Int8)),
+                          ConstantExpr::create(1, Expr::Int8));
```
Comment (domainexpert) [inline lib/Core/SeedInfo.cpp]: Why is the error here 1?

File: lib/Core/SpecialFunctionHandler.cpp [hunk 5]
```diff
@@ -697,18 +697,18 @@ void SpecialFunctionHandler::handleRealloc(ExecutionState &state,
   ref<Expr> address = arguments[0];
   ref<Expr> size = arguments[1];
 
-  Executor::StatePair zeroSize = executor.fork(state, 
-                                               Expr::createIsZero(size), 
-                                               true);
-  
+  Executor::StatePair zeroSize =
+      executor.fork(state, Expr::createIsZero(size),
+                    ConstantExpr::create(1, Expr::Int8), true);
```
Comment (domainexpert) [inline lib/Core/SpecialFunctionHandler.cpp]: Why is the error here 1?

File: lib/Core/Executor.cpp [hunk 6]
```diff
@@ -4557,7 +4568,8 @@ bool Executor::getSymbolicSolution(const ExecutionState &state,
       if (!success) break;
       // If the particular constraint operated on in this iteration through
       // the loop isn't implied then add it to the list of constraints.
-      if (!mustBeTrue) tmp.addConstraint(*pi);
+      if (!mustBeTrue)
+        tmp.addConstraint(*pi, ConstantExpr::create(1, Expr::Int8));
```
Comment (domainexpert) [inline lib/Core/Executor.cpp]: Why is the error here 1?

File: lib/Core/SymbolicError.h [hunk 7]
```diff
@@ -157,6 +161,10 @@ class SymbolicError {
 
   double getPathProbability() const { return pathProbability; }
 
+  void addErrorCondition(ref<Expr> error) {
```
Comment (domainexpert) [inline lib/Core/SymbolicError.h]: @Himeshi I prefer renaming of this to addErrorConstraint for uniformity with the rest of the KLEE code.

**Review by domainexpert — CHANGES_REQUESTED** [review]
Some small changes for now, and please rebase with master. I also requested some explanations in various places. It seems relative error is being used, with the assumption that the relative error is <= 1. However, this may not be true: There can be, for example, 200% error. Please correct me if my question is invalid. Do you have an example that I can use to check that the correct error being generated?

**Comment by domainexpert** [comment]
@Himeshi Do you have an example that you used for testing this feature? Just so that I could test it myself.

**Comment by Himeshi** [comment]
Added a new pull request #56 and therefore closing.

**Pull request closed without merging**